Src/Services.Repo.pas

```diff
@@ -0,0 +1,179 @@
+unit Services.Repo;
+
+interface
+
+uses
+  Services.IRepo,
+
+  Spring.Collections,
+
+  System.SysUtils,
+  System.Variants,
+  System.SyncObjs,
+
+  FireDAC.VCLUI.Wait,
+  FireDAC.Stan.Intf, FireDAC.Stan.Option,
+  FireDAC.Stan.Error, FireDAC.UI.Intf, FireDAC.Stan.Def,
+  FireDAC.Stan.Pool, FireDAC.Stan.Async, FireDAC.Phys,
+  Data.DB, FireDAC.Comp.Client, FireDAC.Phys.SQLite,
+  Firedac.DApt;
+
+type
+
+  TRepo = class(TInterfacedObject, IRepo)
+    private
+      FSQLLite: TFDConnection;
+      FManager: TFDCustomManager;
+      FCT: TCriticalSection;
+
+    public
+      constructor Create;
+
+      function GetFiles(aQuery: string): IList<TSearchResult>;
+      procedure AddFile(aFileName: string; aPath: string);
+      procedure BatchAddFiles(aFiles: IList<string>);
+
+      destructor Destroy; override;
+  end;
+
+
+implementation
+
+uses
+  System.IOUtils;
+
+{ TRepo }
+
+procedure TRepo.AddFile(aFileName, aPath: string);
+begin
+  FCT.Enter;
+  try
+    var lQuery := TFDQuery.Create(nil);
+    lQuery.Connection := FSQLLite;
+    try
+      with lQuery do
+      begin
+        Open('SELECT Path FROM files WHERE Path=?', [aPath]);
+
+        if RecordCount = 0 then
+        begin
+          SQL.Clear;
+          SQL.Add('INSERT INTO files (Path,Name) VALUES (:path,:name)');
+          Params.ParamByName('path').AsString := aPath;
+          Params.ParamByName('name').AsString := aFileName;
+
+          Execute;
+        end;
+      end;
+    finally
+      FreeAndNil(lQuery);
+    end;
+  finally
+    FCT.Leave;
+  end;
+end;
+
+procedure TRepo.BatchAddFiles(aFiles: IList<string>);
+begin
+  FCT.Enter;
+  try
+    var lQuery := TFDQuery.Create(nil);
+    lQuery.Connection := FSQLLite;
+    try
+      with lQuery do
+      begin
+        SQL.Add('delete from files;');
+
+        for var lFile in aFiles do
+        begin
+          var lFileName := TPath.GetFileName(lFile);
+          SQL.Add('INSERT INTO files (Path,Name) VALUES ('''+lFIle+''','''+lFileName+''');');
+        end;
```

Comment:
So I am not quite sure what this `SQL.Add` statement does, but from having worked a lot with SQLite in the pas years, I know from experience, that it is significantly faster to do all the inserts within a transaction so the absolute best statement would look like:
So constructing the sql statement in the loop and then adding it within a transaction is much faster than individual writes.
```sql
BEGIN TRANSACTION;
INSERT INTO files (Path, Name) VALUES ('a', 'b');
INSERT INTO files (Path, Name) VALUES ('c', 'd');
-- ...
COMMIT;
```

Reply:
The SQL.Add adds it all to a list, and then the execute after having it added all, means that its running in a transaction.
So the final SQL that get run looks like what you wrote :)